src/ctap/storage.rs

```diff
@@ -570,6 +570,38 @@ impl PersistentStore {
         Ok(self.store.insert(key::AAGUID, aaguid)?)
     }
 
+    /// Returns the upgrade public key if defined for the identifier.
+    pub fn _upgrade_public_key(
+        &self,
+        identifier: UpgradeIdentifier,
+    ) -> Result<Option<CoseKey>, Ctap2StatusCode> {
+        let value = match identifier {
+            UpgradeIdentifier::A => self.store.find(key::_UPGRADE_PUBLIC_KEY_A)?,
+            UpgradeIdentifier::B => self.store.find(key::_UPGRADE_PUBLIC_KEY_B)?,
+        };
+        value
+            .map(|k| {
+                _deserialize_cose_key(&k).ok_or(Ctap2StatusCode::CTAP2_ERR_VENDOR_INTERNAL_ERROR)
+            })
+            .transpose()
+    }
+
+    /// Sets an upgrade public key.
+    ///
+    /// Overwrites an existing key for the given identifier.
+    pub fn _set_upgrade_public_key(
```

Comment:
I was thinking that client's just configure twice. But aren't those independent questions?
- 1 or 2 persistent keys
- storage API gets and sets 1 or 2 keys

I think for persistent keys, having 2 is better since we can't easily predict the length of their storage representation, so we'd need a CBOR map. CBOR maps need destructuring, and that costs binary size.

I don't have a use case in mind where the getter needs to return both public keys.

Your point with the setter is fair, you will probably always want both configured. I think that also depends on how we will deal with device lock down, and if we allow lockdown without upgradability (for users that rather want to reduce attack surface). One other use case I can think of outside of configure will be when a future firmware rotates the public keys, i.e. when the associated private keys are too old or compromised. That will probably also overwrite both.

Reply:
I see, the keys are not constant length. But why use a map instead of an array? I'm not sure the binary size should influence too much the design (in particular for things like persistent storage key partition that can't be easily changed), but it's indeed a compromise.

Reply:
Or an array, yes. But the cleanest design would probably be to map `UpgradeIdentifier` to public key. That was my first implementation, and I saved around 1 kB with the current approach IIRC. And since in our case a few kB may mean that OpenSK deploys on a certain hardware, I'd give it more weight.

Reply:
1kB is due to some rounding? that's huge. I guess the cbor map implementations are quite big. I don't think arrays would be that bad. I'm not sure a map is cleaner because the fact that we have exactly 2 sides A/B is hard-coded in the whole upgrade design.